src/plugin_tc.c

```diff
@@ -184,399 +187,389 @@ static inline void tc_device_classes_cleanup(struct tc_device *d) {
+        snprintfz(var_name, CONFIG_MAX_NAME, "ctokens chart for %s", d->id);
+        d->enabled_ctokens            = (char)config_get_boolean_ondemand("plugin:tc", var_name, enable_ctokens);
+
+        snprintfz(var_name, CONFIG_MAX_NAME, "show all classes for %s", d->id);
+        d->enabled_all_classes_qdiscs = (char)config_get_boolean_ondemand("plugin:tc", var_name, enabled_all_classes_qdiscs);
     }
 
     // we only need to add leaf classes
-    struct tc_class *c, *x;
+    struct tc_class *c, *x /*, *root = NULL */;
     unsigned long long bytes_sum = 0, packets_sum = 0, dropped_sum = 0, tokens_sum = 0, ctokens_sum = 0;
-    int active_classes = 0;
+    int active_nodes = 0, updated_classes = 0, updated_qdiscs = 0;
+
+    // prepare all classes
+    // we set reasonable defaults for the rest of the code below
 
-    // set all classes
     for(c = d->classes ; c ; c = c->next) {
-        c->isleaf = 1;
-        c->hasparent = 0;
+        c->render = 0;          // do not render this class
+
+        c->isleaf = 1;          // this is a leaf class
+        c->hasparent = 0;       // without a parent
 
         if(unlikely(!c->updated))
-            c->unupdated++;
-        else
-            c->unupdated = 0;
+            c->unupdated++;     // increase its unupdated counter
+        else {
+            c->unupdated = 0;   // reset its unupdated counter
+
+            // count how many of each kind
+            if(c->isqdisc)
+                updated_qdiscs++;
+            else
+                updated_classes++;
+        }
+    }
+
+    if(unlikely(!d->enabled || (!updated_classes && !updated_qdiscs))) {
+        debug(D_TC_LOOP, "TC: Ignoring TC device '%s'. It is not enabled/updated.", d->name?d->name:d->id);
+        tc_device_classes_cleanup(d);
+        return;
+    }
+
+    if(unlikely(updated_classes && updated_qdiscs)) {
+        error("TC: device '%s' has active both classes (%d) and qdiscs (%d). Will render only qdiscs.", d->id, updated_classes, updated_qdiscs);
+
+        // set all classes to !updated
+        for(c = d->classes ; c ; c = c->next)
+            if(unlikely(!c->isqdisc && c->updated))
+                c->updated = 0;
+
+        updated_classes = 0;
     }
 
     // mark the classes as leafs and parents
-    for(c = d->classes ; c ; c = c->next) {
+    //
+    // TC is hierarchical:
+    //  - classes can have other classes in them
+    //  - the same is true for qdiscs (i.e. qdiscs have classes, that have other qdiscs)
+    //
+    // we need to present a chart with leaf nodes only, so that the sum
+    // of all dimensions of the chart, will be the total utilization
+    // of the interface.
+    //
+    // here we try to find the ones we need to report
+    // by default all nodes are marked with: isleaf = 1 (see above)
+    //
+    // so, here we remove the isleaf flag from nodes in the middle
+    // and we add the hasparent flag to leaf nodes we found their parent
+    for(c = d->classes; c; c = c->next) {
```

Comment:
Checking for leaf and parent nodes is not necessary, if `enabled_all_classes_qdiscs` is true. An if statement could be put around this nested loop. This could improve the performance.